compiler/dyno/lib/uast/chpl-syntax-printer.cpp

```diff
@@ -0,0 +1,1065 @@
+        ss_ <<  " = ";
+        printAst(ss_, ie);
+      }
+    }
+
+    //erroneousExpression
+
+    void visit(const ExternBlock* node) {
+      ss_ << "extern {\n";
+      ss_ << node->code();
+      ss_ << "}";
+    }
+
+    void visit(const FnCall* node) {
+      const Expression* callee = node->calledExpression();
+      if (!callee) {
+        printf("ERROR %s\n", asttags::tagToString(node->tag()));
+        node->stringify(std::cerr, StringifyKind::CHPL_SYNTAX);
+      }
+      assert(callee);  // This should be true because OpCall is handled
+
+      printAst(ss_, callee);
+      if (callee->isIdentifier() && (callee->toIdentifier()->name() == "borrowed" || callee->toIdentifier()->name() == "owned"
+                                     || callee->toIdentifier()->name() == "unmanaged" || callee->toIdentifier()->name() == "shared")) {
+        ss_ << " ";
+        printAst(ss_, node->actual(0));
+      } else {
+        if (node->callUsedSquareBrackets()) {
+          interpose(ss_, node->actuals(), ", ", "[", "]");
+        } else {
+          interpose(ss_, node->actuals(), ", ", "(", ")");
+        }
+      }
+
+    }
+
+    void visit(const For* node) {
+      ss_ << "for ";
+      if (node->isParam()) {
+        ss_ << "param ";
+      }
+      if (node->index()) {
+        printAst(ss_, node->index());
+        ss_ << " in ";
+      }
+      printAst(ss_, node->iterand());
+      ss_ << " ";
+      printBlockWithStyle(ss_, node->blockStyle(), node->stmts(), "do ");
+    }
+
+    void visit(const Forall* node) {
+      ss_ << "forall ";
+
+      if (node->index()) {
+        printAst(ss_, node->index());
+        ss_ << " in ";
+      }
+      printAst(ss_, node->iterand());
+      if (node->withClause()) {
+        ss_<< " ";
+        printAst(ss_, node->withClause());
+      }
+      ss_ << " ";
+      printBlockWithStyle(ss_, node->blockStyle(), node->stmts(), "do ");
+    }
+
+    void visit(const Foreach* node) {
+      ss_ << "foreach ";
+
+      if (node->index()) {
+        printAst(ss_, node->index());
+        ss_ << " in ";
+      }
+      printAst(ss_, node->iterand());
+      ss_ << " ";
+      printBlockWithStyle(ss_, node->blockStyle(), node->stmts(), "do ");
+    }
+
+    void visit(const Formal* node) {
+      if (node->intent() != Formal::DEFAULT_INTENT) {
+        ss_ << kindToString(node->intent()) << " ";
+      }
+      ss_ << node->name().str();
+      if (const Expression* te = node->typeExpression()) {
+        ss_ << ": ";
+        printAst(ss_, te);
+      }
+      if (const Expression* ie = node->initExpression()) {
+        ss_ << " = ";
+        printAst(ss_, ie);
+      }
+    }
+
+    void visit(const ForwardingDecl* node) {
+      ss_ << "forwarding ";
+      if (node->expr()) {
+        printAst(ss_, node->expr());
+      }
+    }
+
+    void visit(const Function* node) {
+
+      printLinkages(node);
+
+      if (node->visibility() != Function::Visibility::DEFAULT_VISIBILITY) {
```

Comment:
Think the visibility might be before the linkage, double check the grammar?

Reply:
I think for `extern` and `export` they are first, and there would be no visibility for them.

```
procedure-declaration-statement:
  privacy-specifier[OPT] procedure-kind[OPT] 'proc' identifier argument-list[OPT] return-intent[OPT] return-type[OPT] where-clause[OPT] function-body
  privacy-specifier[OPT] procedure-kind[OPT] 'operator' operator-name argument-list return-intent[OPT] return-type[OPT] where-clause[OPT] function-body

external-procedure-declaration-statement:
  'extern' external-name[OPT] 'proc' identifier argument-list return-intent[OPT] return-type[OPT]

exported-procedure-declaration-statement:
  'export' external-name[OPT] 'proc' identifier argument-list return-intent[OPT] return-type[OPT]
    function-body
```

Reply:
Verified in `chpl.ypp` that visibility doesn't apply to `extern` or `export` procedures.